Assets/MRTK/SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs

```diff
@@ -114,54 +157,168 @@ protected override bool IsValidController(IMixedRealityController controller)
                         {
-                            RotationBehavior = SolverRotationBehavior.LookAtTrackedObject;
-                            OffsetBehavior = SolverOffsetBehavior.TrackedObjectRotation;
-                        }
-                        // If we are within the threshold angle, we snap to follow the camera
-                        else
-                        {
-                            RotationBehavior = SolverRotationBehavior.LookAtMainCamera;
-                            OffsetBehavior = SolverOffsetBehavior.LookAtCameraRotation;
+                            return IsUserGazingAtActivationPoint(jointedHand);
                         }
                     }
+
+                    eyeGazeActivationAlreadyTriggered = palmFacingThresholdMet ? eyeGazeActivationAlreadyTriggered : false;
+
+                    return palmFacingThresholdMet;
                 }
                 else
                 {
                     Debug.LogError("HandConstraintPalmUp requires controllers of type IMixedRealityHand to perform hand activation tests.");
                 }
 
+                eyeGazeActivationAlreadyTriggered = palmFacingThresholdMet ? eyeGazeActivationAlreadyTriggered : false;
                 return palmFacingThresholdMet;
             }
 
             return true;
         }
+
+        /// <summary>
+        /// Checks to see if the palm is currently facing the user
+        /// </summary>
+        /// <param name="jointedHand"></param>
+        /// <param name="palmPose"></param>
+        /// <param name="palmCameraAngle"></param>
+        /// <returns></returns>
+        private bool IsPalmFacingUpwards(IMixedRealityHand jointedHand, MixedRealityPose palmPose, float palmCameraAngle)
+        {
+            if (requireFlatHand)
+            {
+                // Check if the triangle's normal formed from the palm, to index, to ring finger tip roughly matches the palm normal.
+                MixedRealityPose indexTipPose, ringTipPose;
+
+                if (jointedHand.TryGetJoint(TrackedHandJoint.IndexTip, out indexTipPose) &&
+                    jointedHand.TryGetJoint(TrackedHandJoint.RingTip, out ringTipPose))
+                {
+                    var handNormal = Vector3.Cross(indexTipPose.Position - palmPose.Position,
+                                                   ringTipPose.Position - indexTipPose.Position).normalized;
+                    handNormal *= (jointedHand.ControllerHandedness == Handedness.Right) ? 1.0f : -1.0f;
+
+                    if (Vector3.Angle(palmPose.Up, handNormal) > flatHandThreshold)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // Check if the palm angle meets the prescribed threshold
+            return palmCameraAngle < facingCameraTrackingThreshold;
+        }
+
+        /// <summary>
+        /// Checks to see if the user is currently gazing at the activation point
```

Comment:
What happens if eye gaze is not available? Does it use the head gaze ray? Would be good to describe that in the method comment.

Reply:
It does revert to that. I'll add it into the method comment